bitsandbytes/backends/base.py

```diff
@@ -0,0 +1,133 @@
+from abc import ABC, abstractmethod
+from typing import Optional, Tuple
+
+import torch
+
+from bitsandbytes.utils import QuantState
+
+
+class Backend(ABC):
+    """Base class for devices backends that will implement their own 8bits and 4bits functions."""
+
+    @abstractmethod
+    def double_quant(
+        self,
+        A,
+        col_stats=None,
+        row_stats=None,
+        out_col=None,
+        out_row=None,
+        threshold=0.0,
+    ):
+        raise NotImplementedError
+
+    @abstractmethod
+    def transform(
```

Comment:
Only needed for CUDA. This will probably not be needed for any other device. See the discussion above.

Reply:
This was somewhat touched upon https://github.com/TimDettmers/bitsandbytes/pull/898#discussion_r1484691111 – this PR doesn't yet move all of the CUDA-specific things into place, but I think that's fine and we can clean it up in near-future work...